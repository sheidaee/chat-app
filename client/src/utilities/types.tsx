export interface Validate {
  _id?: string;
  name?: string;
  origin?: string;
  destination?: string;
  assignee?: string;
  orderStatus?: string;
  pickupDate?: string | Date | null;
  deliveryDate?: string | Date | null;
}

export interface AuthValidate {
  _id?: string;
  username?: string;
  password?: string;
}
