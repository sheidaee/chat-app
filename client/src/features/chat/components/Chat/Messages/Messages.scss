@import "../../../../../sass/abstracts/mixins";
@import "../../../../../sass/abstracts/theme";

/* thread container */
.threadContainer {
  display: flex;
  justify-content: center;
	flex-grow: 1;
	overflow-y: scroll;
	position: relative;

  &::-webkit-scrollbar-track {
    -webkit-box-shadow: inset 0 0 0.6rem rgba(0, 0, 0, 0.3);
    background-color: #cac8ee;
  }

  &::-webkit-scrollbar {
    width: 0.5rem;
    background-color: green;
  }

  &::-webkit-scrollbar-thumb {
    background-color: #484d79;
  }

  .thread {
    position: relative;
    width: 50%;
    min-height: 80rem;
    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    @include themify($themes) {
      color: themed('threadContainerThreadColor');
    }

    @include respond(tab-land) {
      width: 100%;
    }
  }
}
/* END thread container */

@keyframes show {
	0% {
		opacity: 0;
	}
	100% {
		opacity: 1;
	};
}


/* message container */
.messageContainer {
	display: flex;
	justify-content: flex-start;
	min-height: 5rem;
	margin: 1rem 1.5rem;
	animation: .65s ease-out 0s show;


  .time {
    order: 1;
  }

  .data {
    order: 2;
    height: 100%;
    margin-left: 2.5rem;
  }

  .name {
    margin-top: 0.5rem;
    text-align: right;
  }

  .message {
    @include themify($themes) {
      background: themed('threadContainerMessageBgColor');
      color: themed('threadContainerMessageColor');
    }
    border-radius: 0.5rem;
    border-top-left-radius: 0;
    box-sizing: border-box;
    height: 100%;
    padding: 1rem 1.5rem;
    position: relative;
  }

  .message::before {
    border-bottom-color: transparent;
    border-left-color: transparent;
    @include themify($themes) {
      border-right-color: themed('threadContainerMessageBeforeBorderColor');
      border-top-color: themed('threadContainerMessageBeforeBorderColor');
    }
    border-style: solid;
    border-width: 0.4rem;
    content: '';
    height: 0;
    left: -0.7rem;
    position: absolute;
    top: 0;
    width: 0;
  }

  &.right {
    text-align: right;
    justify-content: flex-end;


    .time {
      order: 2;
      margin-left: 2.5rem;
    }

    .data {
      margin-left: 0;
      order: 1;
    }

    .name {
      display: none;
    }

    .message {
      @include themify($themes) {
        background: themed('threadContainerMessageRightBgColor');
        color: themed('threadContainerMessageRightColor');
      }
      border-top-right-radius: 0;
      border-top-left-radius: 0.5rem;

      &::before {
        @include themify($themes) {
          border-top-color: themed('threadContainerMessageRightBeforeBorderColor');
          border-left-color: themed('threadContainerMessageRightBeforeBorderColor');
        }
        border-right-color: transparent;
        left: auto;
        right: -0.7rem;
      }
    }
  }
}
/* END message container */

.typingUser {
	text-align: right;
	margin: 1rem 1.5rem;
}
