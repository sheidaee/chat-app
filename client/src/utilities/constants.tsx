// General Page Section
export const CHAT_PAGE = '/chat';

export const SETTINGS_PAGE = '/settings';

export const LOGIN_PAGE = '/';

export const SOCKET_URL = 'http://localhost:3231/';

export const interfaceColorItems: any = {
  en: [
    { label: 'Light', value: '1' },
    { label: 'Dark', value: '2' },
  ],
  de: [
    { label: 'Light', value: '1' },
    { label: 'Dark', value: '2' },
  ],
};

export const clockDisplayItems: any = {
  en: [
    { label: '12 Hours', value: '1' },
    { label: '24 Hours', value: '2' },
  ],
  de: [
    { label: '12 Hours', value: '1' },
    { label: '24 Hours', value: '2' },
  ],
};

export const onOffItems: any = {
  en: [
    { label: 'On', value: '1' },
    { label: 'Off', value: '2' },
  ],
  de: [
    { label: 'On', value: '1' },
    { label: 'Off', value: '2' },
  ],
};

export const languageItems = [
  { caption: 'English', value: 'en' },
  { caption: 'Deutsch', value: 'de' },
];
