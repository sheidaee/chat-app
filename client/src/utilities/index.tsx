export { updateObject, checkValidity, classNames } from './utility';
