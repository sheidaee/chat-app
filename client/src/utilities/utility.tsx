import { memo, ReactNode } from 'react';
import { SettingsFormValues } from '../features/chat/types';

/**
 * update old object immutable
 *
 * @param {object} oldObject
 * @param {object} updatedProperties
 * @returns updated object
 */
export function updateObject(oldObject: object, updatedProperties: object) {
  return {
    ...oldObject,
    ...updatedProperties,
  };
}

interface Roles {
  required: boolean;
  minLength?: number;
  maxLength?: number;
  isEmail?: boolean;
  isNumeric?: boolean;
}

/**
 * validating form field
 *
 * @export
 * @param {string|number} value
 * @param {object} rules
 * @returns true if is valid
 */
export function checkValidity(value: any, rules: Roles) {
  let isValid = true;
  if (!rules) {
    return true;
  }

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
  }

  if (rules.isNumeric) {
    const pattern = /^\d+$/;
    isValid = pattern.test(value) && isValid;
  }

  return isValid;
}

/**
 * Joining strings with space
 *
 * @param {array|string} class names
 * @param {array} rest
 *
 * @returns {string} class names
 */
export function classNames(classes: string[] | string, ...rest: string[]) {
  if (classes && classes.constructor === Array) {
    return (classes as string[]).join(' ');
  }
  if (arguments[0] !== undefined) {
    return [...(classes as []), ...rest].join(' ');
  }

  return '';
}

export const validate = (values: SettingsFormValues) => {
  const errors: any = {};

  if (!values.username) errors.username = 'Please enter user name';

  return errors;
};

export const withMemo = (Component: ReactNode, checkedProps: any) => {
  function areEqual(prevProps: any, nextProps: any) {
    let isEqual = true;

    checkedProps.forEach((prop: string) => {
      if (prevProps[prop] !== nextProps[prop]) {
        isEqual = false;
      }
    });
    return isEqual;
  }

  return memo(Component as any, areEqual);
};

function convertTime12to24(time12h: string) {
  const [time, modifier] = time12h.split(' ');

  // eslint-disable-next-line prefer-const
  let [hours, minutes]: any = time.split(':');

  if (hours === '12') {
    hours = '00';
  }

  if (modifier === 'PM') {
    hours = parseInt(hours, 10) + 12;
  }

  return `${hours}:${minutes}`;
}

export function formatTime(str: string, twentyFourHours = false) {
  const date = new Date(str);

  let curHour: any = date.getHours();
  let minutes: any = date.getMinutes();
  let AMorPM: any;

  AMorPM = curHour >= 12 ? (AMorPM = 'PM') : (AMorPM = 'AM');
  curHour = curHour > 12 ? (curHour -= 12) : curHour;

  if (curHour < 10) curHour = `0${curHour}`;
  if (minutes < 10) minutes = `0${minutes}`;

  const finalDate = `${curHour}:${minutes} ${AMorPM}`;

  if (twentyFourHours) {
    return convertTime12to24(finalDate);
  }

  return finalDate;
}
