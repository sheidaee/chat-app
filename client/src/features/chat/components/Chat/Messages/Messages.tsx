import React, { useEffect, useRef } from 'react';
import { useSelector } from 'react-redux';

import useTranslation from '../../../../../hooks/useTranslation';
import { formatTime } from '../../../../../utilities/utility';

import './Messages.scss';

interface Messages {
  id: string;
  time: string;
  message: string;
  sender: string;
}

const Messages = () => {
  const chats = useSelector(({ app }: any) => app.chats);
  const activeChat = useSelector(({ app }: any) => app.activeChat);
  const user = useSelector(({ app }: any) => app.user);
  const users = useSelector(({ app }: any) => app.users);
  const clockDisplay = useSelector(({ app }: any) => app.settings.clockDisplay);
  const { t } = useTranslation();
  const containerRef = useRef() as React.MutableRefObject<HTMLInputElement>;

  useEffect(() => {
    containerRef.current.scrollTop = containerRef.current.scrollHeight;
  });

  const convertToTwentyFourHorse = clockDisplay === '2';

  return (
    <div
      className="threadContainer"
      ref={containerRef}
      data-testid="threadContainer"
    >
      <div className="thread">
        {chats[activeChat] &&
          chats[activeChat].messages.map(
            ({ id, time, message, sender }: Messages) => {
              return (
                <div
                  key={id}
                  className={`messageContainer ${sender === user.id &&
                    'right'}`}
                  data-testid="messageContainer"
                >
                  <div className="time">
                    {formatTime(time, convertToTwentyFourHorse)}
                  </div>
                  <div className="data">
                    <div className="message">{message}</div>
                    <div className="name">{users[sender].name}</div>
                  </div>
                </div>
              );
            }
          )}
        {chats[activeChat].typingUsers.map((userId: string) => (
          <div key={userId} className="typingUser">
            {`${users[userId].name} ${t('isTyping')} . . .`}
          </div>
        ))}
      </div>
    </div>
  );
};

export default Messages;
