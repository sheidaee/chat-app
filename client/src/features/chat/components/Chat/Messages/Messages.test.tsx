import React from 'react';
import { createStore } from 'redux';
import { cleanup } from '@testing-library/react';

import { formatTime } from '../../../../../utilities/utility';
import { renderWithRedux, generate } from 'til-client-test-utils';
import Messages from './Messages';

const newUser = generate.userData();
const user = { id: newUser.id, name: newUser.name };
const users = { [user.id]: user };

afterEach(cleanup);

test('should display Messages', async () => {
  const messages = generate.messagesData({
    count: 1,
    overrideMessage: { sender: user.id },
  });

  const chats: any = {
    community: {
      messages,
      typingUsers: [],
    },
  };

  const app = {
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const store = createStore(() => ({ app }));

  const { getAllByTestId, getByText } = renderWithRedux(<Messages />, {
    store,
  });

  const items = await getAllByTestId('messageContainer');
  const { time, message }: any = messages[0];

  expect(items).toHaveLength(1);
  expect(getByText(message)).toBeTruthy();
  expect(getByText(user.name)).toBeTruthy();
  expect(getByText(formatTime(time))).toBeTruthy();
});

test('should display all messages', async () => {
  const messages = generate.messagesData({
    count: 10,
    overrideMessage: { sender: user.id },
  });

  const chats: any = {
    community: {
      messages,
      typingUsers: [],
    },
  };

  const app = {
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const store = createStore(() => ({ app }));

  const { getAllByTestId } = renderWithRedux(<Messages />, {
    store,
  });

  const items = await getAllByTestId('messageContainer');

  expect(items).toHaveLength(10);
});

test('snapshot', async () => {
  const chats: any = {
    community: {
      messages: [],
      typingUsers: [],
    },
  };

  const app = {
    user,
    users,
    chats,
    activeChat: 'community',
    settings: { clockDisplay: '1' },
  };

  const store = createStore(() => ({ app }));

  const { container } = renderWithRedux(<Messages />, {
    store,
  });

  expect(container).toMatchSnapshot();
});
