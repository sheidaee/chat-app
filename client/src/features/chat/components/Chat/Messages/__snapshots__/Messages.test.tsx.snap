// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <div
    class="threadContainer"
    data-testid="threadContainer"
  >
    <div
      class="thread"
    />
  </div>
</div>
`;
